### PR TITLE
>>> DO NOT MERGE, It's a WIP <<<   HHH-11263 - BootstapContext + HHH-11264 Event Listeners merge 

### PR DESCRIPTION
**Pull request by dreab8**
@sebersole the part related with the `org.hibernate.type.spi.TypeConfiguration` is not finished [Intoduce TypeConfiguration](https://github.com/hibernate/hibernate-orm/pull/2203/commits/8ff11f167ba26e9d1c007d20ec1794de1c637f86) , anyway the code compile and all the test pass.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/boot/internal/MetadataImpl.java [hunk 1]
```diff
@@ -84,7 +85,7 @@
 	public MetadataImpl(
 			UUID uuid,
 			MetadataBuildingOptions metadataBuildingOptions,
-			TypeResolver typeResolver,
+			TypeConfiguration typeConfiguration,
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/MetadataImpl.java]: Is there a reason we can't just get the `TypeConfiguration` from the `BoootstrapContext` here?

File: hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java [hunk 2]
```diff
@@ -311,45 +314,34 @@ public void finishUp() {
 		return metadataCollector.buildMetadataInstance( rootMetadataBuildingContext );
 	}
 
-//	private static JandexInitManager buildJandexInitializer(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java]: Can we leave this in place?  Possibly with a `todo (7.0)` comment?

**Comment by sebersole** [comment]
You may already have done it (its hard to see looking on line) but we should make sure that `TypeResolver` itself and all of its uses are depprecated

File: hibernate-core/src/main/java/org/hibernate/boot/internal/MetadataImpl.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/boot/internal/MetadataImpl.java]: just an oversight. Thanks

File: hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java]: Added back.

**Comment by dreab8** [comment]
@sebersole Thanks for the review and the comments. 
About TypeResolver deprecation,  I have to go through and deprecate all of the uses. 
Should I add all the TypeResolver methods to the TypeConfiguration and then deprecate them in favour of the TypeConfiguration ones?

**Comment by sebersole** [comment]
> Should I add all the TypeResolver methods to the TypeConfiguration and then deprecate them in favour of the TypeConfiguration ones?

No.  I think you did it correctly - expose `TypeConfiguration#getTypeResolver` (deprecated).  Just wanted to make sure we got TypeResolver itself deprecated and any accesses of it (like immediatelly deprecating this `TypeConfiguration#getTypeResolver` method).

File: hibernate-core/src/main/java/org/hibernate/boot/spi/BootstrapContext.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	/**
+	 * Releases the "bootstrap only" resources held by this BootstrapContext.
+	 * <p/>
+	 * Only one call to this method is supported, after we have completed the process of
+	 * building the (non-inflight) Metadata impl.  We may want to delay this until we
+	 * get into SF building.  Not sure yet.
+	 *
+	 * @todo verify this ^^
+	 */
+	void release();
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/spi/BootstrapContext.java]: I'm not seeing where this gets called....

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/boot/spi/BootstrapContext.java]: In 6.0 this happens in `InFlightMetadataCollectorImpl#buildMetadataInstance`

File: hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java [hunk 4]
```diff
@@ -247,8 +253,9 @@ public SessionFactoryImpl(final MetadataImplementor metadata, SessionFactoryOpti
 		for ( SessionFactoryObserver sessionFactoryObserver : options.getSessionFactoryObservers() ) {
 			this.observer.addObserver( sessionFactoryObserver );
 		}
-
-		this.typeResolver = metadata.getTypeResolver().scope( this );
+		TypeConfiguration typeConfiguration = metadata.getTypeConfiguration();
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java]: IMO we should keep the TypeConfiguration either here or (more likely) on MetamodelImpl and access things from there.  E.g., here I would just drop the `typeResolver` field - we can always just get it from TypeConfiguration

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java]: do you mean to add a `getTypeConfiguration() ` method to `MetamodelImplementor`?

**Comment by sebersole** [comment]
Yes.  On 6 we add it to Metamodel, but I think MetamodelImplementor is
probably more appropriate

**Comment by sebersole** [comment]
Actually... wrt MetamodelImplementor#getTypeConfiguration... I think what I'd really like to do instead is to essentially replace `org.hibernate.Metamodel` / `org.hibernate.metamodel.spi.MetamodelImplementor` with TypeConfiguration.. that is the direction I was headed in 6.0.  To that end, I would:

1. Deprecate both `org.hibernate.Metamodel` and `org.hibernate.metamodel.spi.MetamodelImplementor`  pointing to `TypeConfiguration` as the replacement
2. Implement as many of the `org.hibernate.Metamodel` and `org.hibernate.metamodel.spi.MetamodelImplementor`  methods as possible by delegating to `TypeConfiguration

**Comment by dreab8** [comment]
@sebersole  [this commit](https://github.com/hibernate/hibernate-orm/pull/2203/commits/f9835865e6e94ea5f589e86ea0600b6aa611005d) applies the changes related to yours first comments
Now I'm going to address your last comment

**Comment by sebersole** [comment]
Awesome @dreab8 !

**Comment by dreab8** [comment]
superseded by https://github.com/hibernate/hibernate-orm/pull/2225

**Pull request closed without merging**